pretyping/typing.ml

```diff
@@ -49,18 +49,20 @@ let e_assumption_of_judgment env evdref j =
   with Type_errors.TypeError _ | PretypeError _ ->
     error_assumption env !evdref j
 
-let e_judge_of_apply env evdref funj argjv =
+let e_judge_of_applied_inductive_knowing_parameters env evdref funj ind argjv =
   let rec apply_rec n typ = function
   | [] ->
       { uj_val  = mkApp (j_val funj, Array.map j_val argjv);
-        uj_type = typ }
-  | hj::restjl ->
-      match EConstr.kind !evdref (whd_all env !evdref typ) with
+        uj_type = whd_all env !evdref (EConstr.of_constr (snd (Term.decompose_prod_n_assum n (inductive_type_knowing_parameters env !evdref ind argjv)))) }
+  | hj::restjl as allj ->
+      match EConstr.kind !evdref (whd_allnolet env !evdref typ) with
+      | LetIn (_,c1,_,c2) ->
+	  apply_rec (n+1) (subst1 c1 c2) allj
       | Prod (_,c1,c2) ->
 	 if Evarconv.e_cumul env evdref hj.uj_type c1 then
 	   apply_rec (n+1) (subst1 hj.uj_val c2) restjl
 	 else
-	   error_cant_apply_bad_type env !evdref (n, c1, hj.uj_type) funj argjv
+           error_cant_apply_bad_type env !evdref (n+1, c1, hj.uj_type) funj argjv
```

Comment:
Why change `n` to `n+1`?

Reply:
Because I changed the initial value of the loop from 1 to 0, so that `n` now counts the number of binders, but there is indeed a problem which is that the expected `n` is here the number of effective arguments, not the number of binders. I fixed it.